netket/logging/hdf5_log.py

```diff
@@ -0,0 +1,173 @@
+    - arrays are stored in the same way, but with values having shape (n_steps, *array_shape),
+    - netket.stats.Stats are stored as a group containing each field (Mean, Variance, etc...) as a separate dataset.
+
+    Importantly, each group has a dataset `iters`, which tracks the iteration number of the logged quantity.  
+
+    If the model state is serialized, then it is serialized as a dataset in the group `parameters/`.
+    The target of the serialization are the parameters of the variational state,
+    which are stored in a dataset `parameters/value` as a flattened array.
+
+    Data can be deserialized by calling :code:`f = h5py.File(filename, 'r')` and
+    inspecting the datasets as a dictionary, i.e. :code:`f['data/energy/Mean']`
+    """
+
+    def __init__(
+        self,
+        output_prefix: str,
+        mode: str = "write",
+        save_params: bool = True,
+        save_params_every: int = 50
+    ):
+        """
+        Construct a HDF5 Logger.
+
+        Args:
+            output_prefix: the name of the output files before the extension
+            mode: Specify the behaviour in case the file already exists at this
+                output_prefix. Options are
+                - `[w]rite`: (default) overwrites file if it already exists;
+                - `[x]` or `fail`: fails if file already exists;
+            save_params: bool flag indicating whether variables of the variational state
+                should be serialized at some interval
+            save_params_every: every how many iterations should machine parameters be
+                flushed to file
+        """
+        import h5py
+        super().__init__()
+
+        if not ((mode == "write") or (mode == "append") or (mode == "fail")):
+            raise ValueError(
+                "Mode not recognized: should be one of `[w]rite`, `[a]ppend` or"
+                "`[x]`(fail)."
+            )
+        mode = _mode_shorthands[mode]
+
+        file_exists = os.path.exists(output_prefix + ".hdf5")
+
+        if file_exists and mode == "x":
+            raise ValueError(
+                "Output file already exists. Either delete it manually or"
+                "change `output_prefix`."
+            )
+
+        dir_name = os.path.dirname(output_prefix)
+        if dir_name != "":
+            os.makedirs(dir_name, exist_ok=True)
+
+        self._prefix = output_prefix
+        self._file_mode = mode
+        self._file_name = self._prefix+'.hdf5'
+        self._writer = None
+
+        self._save_params = save_params
+        self._save_params_every = save_params_every
+
+    def __call__(self, step, log_data, variational_state):
+        if self._writer is None:
+            self._writer = h5py.File(self._file_name, self._file_mode)
+        tree_log(log_data, "data", self._writer, iter=step)
+        if self._save_params and step % self._save_params_every == 0:
```

Comment:
`step` is an integer when running `VMC` but it's a floating point (the time) when running `TDVP`, so this code won't work.

The way to do this that we used in to keep an internal counter of the steps, `self._steps_notsaved_params` and increase it by 1 at the end of the `__call__` function. See for example [StateLog](https://github.com/netket/netket/blob/6e0d9eef36fe60bc923a399195cfca08f7f31766/netket/logging/state_log.py#L155)

```suggestion
        if self._steps_notsaved_params and step % self._save_params_every == 0:
```